jib-core/src/main/java/com/google/cloud/tools/jib/api/JibContainerBuilder.java

```diff
@@ -417,21 +420,45 @@ public JibContainerBuilder setWorkingDirectory(@Nullable AbsoluteUnixPath workin
   public JibContainer containerize(Containerizer containerizer)
       throws InterruptedException, ExecutionException, IOException,
           CacheDirectoryCreationException {
-    BuildConfiguration buildConfiguration = toBuildConfiguration(containerizer);
-    BuildResult result = containerizer.getTargetImage().toBuildSteps(buildConfiguration).run();
-    return new JibContainer(result.getImageDigest(), result.getImageId());
+
+    return containerize(containerizer, Executors::newCachedThreadPool);
+  }
+
+  @VisibleForTesting
+  JibContainer containerize(
+      Containerizer containerizer, Supplier<ExecutorService> executorServiceFactory)
+      throws InterruptedException, ExecutionException, IOException,
+          CacheDirectoryCreationException {
+
+    boolean shutdownExecutorService = !containerizer.getExecutorService().isPresent();
+    ExecutorService executorService =
+        containerizer.getExecutorService().orElseGet(executorServiceFactory);
+
+    BuildConfiguration buildConfiguration = toBuildConfiguration(containerizer, executorService);
+
+    try {
+      BuildResult result = containerizer.getTargetImage().toBuildSteps(buildConfiguration).run();
+      return new JibContainer(result.getImageDigest(), result.getImageId());
+    } finally {
+      if (shutdownExecutorService) {
```

Comment:
Could we control the lifecycle with an `Consumer<ExecutorService> cleanup` option?

I'm having a hard time visualizing it though, so maybe we shouldn't.

Reply:
I think `try ... finally` works. (I'm assuming the `run()` is blocking, which I believe is true.)

Reply:
The `run()` is somewhat-blocking, in that we [wait on the future from the final step](https://github.com/GoogleContainerTools/jib/blob/master/jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/StepsRunner.java#L276), but `shutdown()` triggers an orderly shutdown that will continue any pending work until completion.

Reply:
> somewhat-blocking,

I'd say it's sure-blocking. :)